src/groupeddataframe/grouping.jl

```diff
@@ -443,71 +612,140 @@ colwise(fns::Union{AbstractVector, Tuple}, d::AbstractDataFrame) = [f(d[i]) for
-               c = randn(8))
-by(df, :a, d -> sum(d[:c]))
-by(df, :a, d -> 2 * skipmissing(d[:c]))
-by(df, :a, d -> (c_sum = sum(d[:c]), c_mean = mean(skipmissing(d[:c]))))
-by(df, :a, d -> DataFrame(c = d[:c], c_mean = mean(skipmissing(d[:c]))))
-by(df, [:a, :b]) do d
-    (m = mean(skipmissing(d[:c])), v = var(skipmissing(d[:c])))
-end
+```jldoctest
+julia> df = DataFrame(a = repeat([1, 2, 3, 4], outer=[2]),
+                      b = repeat([2, 1], outer=[4]),
+                      c = 1:8);
+
+julia> by(df, :a, :c => sum)
+4×2 DataFrame
+│ Row │ a     │ c_sum │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 6     │
+│ 2   │ 2     │ 8     │
+│ 3   │ 3     │ 10    │
+│ 4   │ 4     │ 12    │
+
+julia> by(df, :a, d -> sum(d.c)) # Slower variant
+4×2 DataFrame
+│ Row │ a     │ x1    │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 6     │
+│ 2   │ 2     │ 8     │
+│ 3   │ 3     │ 10    │
+│ 4   │ 4     │ 12    │
+
+julia> by(df, :a) do d # do syntax for the slower variant
+           sum(d.c)
+       end
+4×2 DataFrame
+│ Row │ a     │ x1    │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 6     │
+│ 2   │ 2     │ 8     │
+│ 3   │ 3     │ 10    │
+│ 4   │ 4     │ 12    │
+
+julia> by(df, :a, :c => x -> 2 .* x)
+8×2 DataFrame
+│ Row │ a     │ c_function │
+│     │ Int64 │ Int64      │
+├─────┼───────┼────────────┤
+│ 1   │ 1     │ 2          │
+│ 2   │ 1     │ 10         │
+│ 3   │ 2     │ 4          │
+│ 4   │ 2     │ 12         │
+│ 5   │ 3     │ 6          │
+│ 6   │ 3     │ 14         │
+│ 7   │ 4     │ 8          │
+│ 8   │ 4     │ 16         │
+
+julia> by(df, :a, c_sum = :c => sum, c_sum2 = :c => x -> sum(x.^2))
+4×3 DataFrame
+│ Row │ a     │ c_sum │ c_sum2 │
+│     │ Int64 │ Int64 │ Int64  │
+├─────┼───────┼───────┼────────┤
+│ 1   │ 1     │ 6     │ 26     │
+│ 2   │ 2     │ 8     │ 40     │
+│ 3   │ 3     │ 10    │ 58     │
+│ 4   │ 4     │ 12    │ 80     │
+
+julia> by(df, :a, (:b, :c) => x -> (minb = minimum(x.b), sumc = sum(x.c)))
+4×3 DataFrame
+│ Row │ a     │ minb  │ sumc  │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 6     │
+│ 2   │ 2     │ 1     │ 8     │
+│ 3   │ 3     │ 2     │ 10    │
+│ 4   │ 4     │ 1     │ 12    │
 ```
 
 """
-by(d::AbstractDataFrame, cols, f::Function; sort::Bool = false) =
+by(d::AbstractDataFrame, cols::Any, f::Any; sort::Bool = false) =
     combine(f, groupby(d, cols, sort = sort))
-by(f::Function, d::AbstractDataFrame, cols; sort::Bool = false) =
+by(f::Any, d::AbstractDataFrame, cols::Any; sort::Bool = false) =
     by(d, cols, f, sort = sort)
+by(d::AbstractDataFrame, cols::Any, f::Pair; sort::Bool = false) =
+    combine(f, groupby(d, cols, sort = sort))
+by(d::AbstractDataFrame, cols::Any, f::Pair...; sort::Bool = false) =
+    combine(f, groupby(d, cols, sort = sort))
+by(d::AbstractDataFrame, cols::Any; sort::Bool = false, f...) =
+    combine(values(f), groupby(d, cols, sort = sort))
```

Comment:
In this section of code I can see why you have implemented `combine` differently than `by` but, as stated above, I feel it would be better if all the funtions had the same user-facing API.